packages/studio-be/src/core/migration/bot-migration-service.ts

```diff
@@ -68,58 +58,66 @@ export class BotMigrationService {
-      const filtered = migrations.filter(x => x.type === type)
+  private writeLabel(migrations: number, botId: string, version: string) {
+    const configLabel = `(${version} => ${this.migService.targetVersion})`
+    const changesLabel = `${migrations} change${migrations === 1 ? '' : 's'}`
 
-      if (filtered.length) {
-        filtered.map(x => logger.warn(`- ${x.description}`))
-      } else {
-        logger.warn('- None')
-      }
-    })
+    return chalk`{bold Migration${migrations === 1 ? '' : 's'} Required for ${botId}} ${configLabel} - ${changesLabel}`
   }
 
-  private async executeBotMigrations(botId: string, missingMigrations: MigrationFile[]) {
-    this.logger.info(chalk`
-${_.repeat(' ', 9)}========================================
-{bold ${centerText(
-      `Executing ${missingMigrations.length} migration${missingMigrations.length === 1 ? '' : 's'}`,
-      40,
-      9
-    )}}
-${_.repeat(' ', 9)}========================================`)
-
-    const opts = await this.migService.getMigrationOpts({ botId })
+  private async executeBotMigrations(botId: string, migrations: MigrationFile[], configVersion: string) {
+    const botConfig = await this.configProvider.getBotConfig(botId)
+    const opts = await this.migService.getMigrationOpts({ botId, botConfig })
     let hasFailures = false
 
-    await Promise.mapSeries(missingMigrations, async ({ filename }) => {
-      const result = await this.migService.loadedMigrations[filename].up(opts)
-      debug.forBot(botId, 'Migration step finished', { filename, result })
-      if (result.success) {
-        this.logger.forBot(botId).info(`- ${result.message || 'Success'}`)
-      } else {
-        hasFailures = true
-        this.logger.forBot(botId).error(`- ${result.message || 'Failure'}`)
+    const logger = this.logger.forBot(botId)
+
+    logger.warn(this.writeLabel(migrations.length, botId, configVersion))
+
+    await Promise.mapSeries(migrations, async ({ filename, info }) => {
+      const label = `${types[info.type]} - ${info.description}`
+
+      try {
+        const result = await this.migService.loadedMigrations[filename].up(opts)
+        debug.forBot(botId, 'Migration step finished', { filename, result })
+
+        if (result.success) {
+          logger.warn(`${chalk.green('[success]')} ${label}`)
+        } else if (result.hasChanges) {
+          logger.error(`${chalk.red('[failure]')} ${label}: ${result.message || ''}`)
+        }
+
+        return
+      } catch (err) {
+        logger.attachError(err).error(`${chalk.red('[failure]')} ${label}`)
       }
+
+      hasFailures = true
     })
 
+    logger.warn('')
     if (hasFailures) {
-      return this.logger.error('Could not complete bot migration. It may behave unexpectedly.')
+      return this.logger.error(`[${botId}] Could not complete bot migration. It may behave unexpectedly.`)
     }
 
     await this.configProvider.mergeBotConfig(botId, { version: this.migService.targetVersion })
```

Comment:
Here we only update the bot version if a migration failed. Otherwise, we return on line 108.

Reply:
Nop, the return of line 108 exits the mapSeries, that was a weird way to avoid adding hasFailures twice, but yeah it's not so clear, i've changed it